crates/pep508-rs/src/marker.rs

```diff
@@ -556,9 +556,12 @@ pub struct MarkerExpression {
 
 impl MarkerExpression {
     /// Evaluate a <`marker_value`> <`marker_op`> <`marker_value`> expression
+    ///
+    /// When `env` is `None`, all expressions that reference the environment
+    /// will evaluate as `true`.
     fn evaluate(
         &self,
-        env: &MarkerEnvironment,
+        env: Option<&MarkerEnvironment>,
         extras: &[ExtraName],
```

Comment:
Is the key thing here that we _do_ need to evaluate extras? Otherwise, I could imagine just doing something like `env.map(...).unwrap_or(true)` instead of encoding this in the API.

Reply:
Yes. We specifically want to evaluate extras.

Reply:
I believe the basic issue here is that our [optional dependencies manifest as addition's to a requirement's markers](https://github.com/astral-sh/uv/blob/d3c72896f9a344296b594ad3b0f2c945430a365e/crates/pypi-types/src/metadata.rs#L224-L235). So when we do resolution, we very specifically want to evaluate the "extra" marker expressions because we're _only_ looking for universality with respect to platform ("marker environment"), and _not_ with respect to the set of possible optional dependencies.

In my prototype, I [dropped optional dependencies altogether](https://github.com/astral-sh/uv/blob/7f2b401260e6231bebb867c034b27d2955210fe0/crates/pypi-types/src/metadata.rs#L246-L257). I couldn't move forward without that because it was too easy for, e.g., packages to bring in huge trees of dev dependencies.